Microservices guide general update to 3.1 (#16202)
dotnet/docs: docs/architecture/microservices/net-core-net-framework-containers/net-framework-container-scenarios.md

While .NET Core offers significant benefits for new applications and application patterns, .NET Framework will continue to be a good choice for many existing scenarios.

> However, the recent release of [.NET Core 3.0](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0/), and even more recent [3.1 release](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-1/), with Long Term Support (LTS). includes significant performance improvements that can help justify a migration. The Windows Desktop related features don't really have an impact on the microservices applications.

Author: ok.

## Using third-party .NET libraries or NuGet packages not available for .NET Core

> Third-party libraries are quickly embracing the [.NET Standard](../../../standard/net-standard.md), which enables code sharing across all .NET flavors, including .NET Core. With the .NET Standard Library 2.0 and beyond the API surface compatibility across different frameworks has become significantly larger and, from .NET Core 2.x applications, can also directly reference existing .NET Framework libraries (see [.NET Framework 4.6.1 supporting .NET Standard 2.0](https://github.com/dotnet/standard/blob/master/docs/planning/netstandard-2.0/README.md#net-framework-461-supporting-net-standard-20)).

Reviewer: ```suggestion
Third-party libraries are quickly embracing [.NET Standard](../../../standard/net-standard.md), which enables code sharing across all .NET flavors, including .NET Core. With .NET Standard 2.0 and later, the API surface compatibility across different frameworks has become significantly larger. And .NET Core 2.x applications can also directly reference existing .NET Framework libraries (see [.NET Framework 4.6.1 supporting .NET Standard 2.0](https://github.com/dotnet/standard/blob/master/docs/planning/netstandard-2.0/README.md#net-framework-461-supporting-net-standard-20)).
```

> 

Reviewer: The previous paragraph is still talking about 2.x not 3.1

Author: That's correct here, it said "from .NET Core 2.0" will change to ".NET Core 2.x and later applications"

Author: Fixed.

> In addition, the [Windows Compatibility Pack](../../../core/porting/windows-compat-pack.md) was released on NOV-2017 to extend the API surface available for .NET Standard 2.0 on Windows. This pack allows recompiling most existing code to .NET Standard 2.x with little or no modification, to run on Windows.

Reviewer: The last release was actually two months ago

```suggestion
In addition, the [Windows Compatibility Pack](../../../core/porting/windows-compat-pack.md) extends the API surface available for .NET Standard 2.0 on Windows. This pack allows recompiling most existing code to .NET Standard 2.x with little or no modification, to run on Windows.
```

Author: That's right, NOV-2017 was the first release.

However, even with that exceptional progression since .NET Standard 2.0 and .NET Core 2.1, there might be cases where certain NuGet packages need Windows to run and might not support .NET Core. If those packages are critical for your application, then you will need to use .NET Framework on Windows Containers.

## Using .NET technologies not available for .NET Core

Some .NET Framework technologies are not available in the current version of .NET Core (version 3.1 as of this writing). Some of them will be available in later .NET Core releases, but others do not apply to the new application patterns targeted by .NET Core and might never be available.

> The following list shows most of the technologies that are not available in .NET Core 2.x:

Reviewer: You should probably review 2.x references throughout

```suggestion
The following list shows most of the technologies that are not available in .NET Core 3.1:
```

Author: Yep, that one slipped through!

In addition, the [Windows Compatibility Pack](../../../core/porting/windows-compat-pack.md) extends the API surface available for .NET Standard 2.0 on Windows. This pack allows recompiling most existing code to .NET Standard 2.x with little or no modification, to run on Windows.

However, even with that exceptional progression since .NET Standard 2.0 and .NET Core 2.1, there might be cases where certain NuGet packages need Windows to run and might not support .NET Core. If those packages are critical for your application, then you will need to use .NET Framework on Windows Containers.

## Using .NET technologies not available for .NET Core

Some .NET Framework technologies are not available in the current version of .NET Core (version 3.1 as of this writing). Some of them might be available in later .NET 5 releases, but others don't fit the new application patterns targeted by .NET Core and might never be available.

The following list shows most of the technologies that are not available in .NET Core 3.1:

- ASP.NET Web Forms. This technology is only available on .NET Framework. Currently there are no plans to bring ASP.NET Web Forms to .NET Core.

> - WCF services. Even when a [WCF-Client library](https://github.com/dotnet/wcf) is available to consume WCF services from .NET Core, as of Feb-2020, the WCF server implementation is only available on .NET Framework. This scenario might be considered for future releases of .NET Core, there are even some APIs considered for inclusion in the [Windows Compatibility Pack](../../../core/porting/windows-compat-pack.md).

Author: Sure not now, I prefer to leave it like that, b/c it might change and it's easier to spot, with a date attached, as something that has to be reviewed.

- Workflow-related services. Windows Workflow Foundation (WF), Workflow Services (WCF + WF in a single service), and WCF Data Services (formerly known as ADO.NET Data Services) are only available on .NET Framework. There are currently no plans to bring them to .NET Core.

In addition to the technologies listed in the official [.NET Core roadmap](https://github.com/dotnet/core/blob/master/roadmap.md), other features might be ported to .NET Core or the new [unified .NET platform](https://devblogs.microsoft.com/dotnet/introducing-net-5/). You might consider participating in the discussions on GitHub so that your voice can be heard. And if you think something is missing, please [file a new issue in the Runtime repository](https://github.com/dotnet/runtime/issues/new).

> 

Author: I'd say the technologies are 

![image](https://user-images.githubusercontent.com/7357806/74467138-6da4de80-4e90-11ea-83c1-8a978d00e7b9.png)

Not sure if it's correct to refer to those as technologies 🤔

## Using a platform or API that doesn't support .NET Core

> Some Microsoft and third-party platforms don't support .NET Core. For example, some Azure services provide an SDK that isn't yet available for consumption on .NET Core. This is temporary, because all Azure services will eventually use .NET Core/Standard. You can see the available Azure SDKs in the [Azure SDK Latest Releases](https://azure.github.io/azure-sdk/releases/latest/index.html) page.

Author: Yep, perhaps that was the idea back then but I think it's fair to change this to something not so absolute, because anything can happen, like considering some tech is no longer relevant.

> Some Microsoft and third-party platforms don't support .NET Core. For example, some Azure services provide an SDK that isn't yet available for consumption on .NET Core. This is temporary, because all Azure services will eventually use .NET Core/Standard. You can see the available Azure SDKs in the [Azure SDK Latest Releases](https://azure.github.io/azure-sdk/releases/latest/index.html) page.

Author: Changed.

> 

Reviewer: line 32 mentions 2017. Should probably be updated.

Author: Yep, should be "since"

Author: Updated to as of Feb-2020

> In addition to the technologies listed in the official [.NET Core roadmap](https://github.com/dotnet/aspnetcore/wiki/Roadmap), other features might be ported to .NET Core. For a full list, look at the items tagged as [port-to-core](https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+label%3Aport-to-core) on the CoreFX GitHub site. Note that this list does not represent a commitment from Microsoft to bring those components to .NET Core — the items simply capture requests from the community. If you care about any of the components listed above, consider participating in the discussions on GitHub so that your voice can be heard. And if you think something is missing, please [file a new issue in the Runtime repository](https://github.com/dotnet/runtime/issues/new).

Reviewer: ```suggestion
In addition to the technologies listed in the official [.NET Core roadmap](https://github.com/dotnet/core/blob/master/roadmap.md), other features might be ported to .NET Core. For a full list, look at the items tagged as [port-to-core](https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+label%3Aport-to-core) on the CoreFX GitHub site. Note that this list does not represent a commitment from Microsoft to bring those components to .NET Core — the items simply capture requests from the community. If you care about any of the components listed above, consider participating in the discussions on GitHub so that your voice can be heard. And if you think something is missing, please [file a new issue in the Runtime repository](https://github.com/dotnet/runtime/issues/new).
```

> 

Reviewer: Actually there are no open port-to-core open issues. Revisit the paragraph?

Author: Tweaked and removed!

> ## Using a platform or API that does not support .NET Core

Reviewer: ```suggestion
## Using a platform or API that doesn't support .NET Core
```

> 

Reviewer: On the following paragraph, the package mentioned is already on version 2.9.3 launched 5 days ago. The 2016 right now seems outdated. I'd rewrite that.

Author: You're right, the 2016 is not important now that "it is now generally available (GA) as a stable version"

Author: Removed and added reference to the latest Azure SDK Releases page.

> Some Microsoft or third-party platforms do not support .NET Core. For example, some Azure services provide an SDK that is not yet available for consumption on .NET Core. This is temporary, because all Azure services will eventually use .NET Core. For example, the [Azure DocumentDB SDK for .NET Core](https://www.nuget.org/packages/Microsoft.Azure.DocumentDB.Core) was released as a preview on November 16, 2016, but it is now generally available (GA) as a stable version.

Reviewer: ```suggestion
Some Microsoft and third-party platforms don't support .NET Core. For example, some Azure services provide an SDK that isn't yet available for consumption on .NET Core. This is temporary, because all Azure services will eventually use .NET Core. For example, the [Azure DocumentDB SDK for .NET Core](https://www.nuget.org/packages/Microsoft.Azure.DocumentDB.Core) was released as a preview on November 16, 2016, but it is now generally available (GA) as a stable version.
```
